### PR TITLE
feat: expand custom field solution to support dynamic custom field types

### LINKED ISSUE #307: [Feature request]: Custom fields
### Contact Details

jurre@whale-agency.com

### Is your feature request related to a problem? Please describe?

Based on multiple user requests wishing for the option to add custom fields in order to realise a customised workflow suited to their needs.

### Describe the solution you'd like

Functionality for managing custom fields.

WIP designs: https://www.figma.com/file/2XkiKJ4II7Mux6JJL9hYTm/Shelf-App-%7C-Production-Ready?type=design&node-id=6683-336967&mode=design

- Under settings, add a tab called 'Custom fields'
- Create 'Custom fields' index
- Option to create new 'Custom field' with field type 'Single line text' (more field types will follow)
- Option to delete 'Custom field' when no asset was ever created with that specific custom field active. @DonKoko  Would be good to discuss this in detail. Want to avoid data being lost.
- Option to archive 'Custom field'
- Option to either show or hide the data filled for the custom field on asset page
- Created 'Custom fields' that are active should show in the 'Create/Edit asset' view

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### LINKED ISSUE #362: [Feature request]: Additional custom fields
### Contact Details

_No response_

### Is your feature request related to a problem? Please describe?

We have released a custom field area, we currently have one field. The type is 'Text'

Screenshot 2023-09-15 at 13 34 18

We want to create more custom field types:

**Most valuable**
- Select (Single select / AKA: Dropdown)
- Boolean
- Date (Check https://ui.shadcn.com/ for a component as this is integrated)

**Still valuable but slightly less**
- Multi line text

### Describe the solution you'd like

Add custom field types and validations on our custom field area (under settings)

We have designs ready. 

https://www.figma.com/file/2XkiKJ4II7Mux6JJL9hYTm/Shelf-App-%7C-Production-Ready?type=design&node-id=7291-376406&mode=design

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mahendraHegde**
closes #362 

/claim #362

sample screenshot of an asset having all custom fields

Screenshot 2023-09-20 at 1 12 18 PM

### CODE REVIEW

File: app/components/assets/custom-fields-inputs.tsx [hunk 1]
```diff
@@ -19,13 +29,101 @@ export default function AssetCustomFields({
...
   /** Get the custom fields from the loader */
   const { customFields } = useLoaderData();
+  const { asset } =
+    useLoaderData<typeof loader>() || {}
 
-  const { customFields: customFieldsValues } =
-    useLoaderData<typeof loader>()?.asset || [];
+  const customFieldsValues = asset?.customFields as unknown as ShelfAssetCustomFieldValueType[] || []
+
+  const [dateObj, steDateObj] = useState(customFieldsValues.filter(v => v.value.valueDate).reduce((res, cur) => {
```
Comment (DonKoko) [inline app/components/assets/custom-fields-inputs.tsx]: Small spelling mistake. Should be `setDateObj`

File: app/database/migrations/20230919162512_expand_custom_fields/migration.sql [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+--- rename valueJson to value
+ALTER TABLE "AssetCustomFieldValue"
+RENAME COLUMN "valueJson" TO "value";
+
+--make it nonNull
+ALTER TABLE "AssetCustomFieldValue"
+ALTER COLUMN "value" SET NOT NULL;
+
+-- add constaint to json schema
+ALTER TABLE "AssetCustomFieldValue"
```
Comment (DonKoko) [inline app/database/migrations/20230919162512_expand_custom_fields/migration.sql:47]: Can you explain that a bit further. I would like to understand exactly what we are doing here.

File: app/routes/_layout+/assets.$assetId.tsx [hunk 3]
```diff
@@ -309,7 +311,7 @@ export default function AssetDetailsPage() {
                       <span className="text-[12px] font-medium text-gray-600">
                         {field.customField.name}
                       </span>
-                      <div className="max-w-[250px]">{field.value}</div>
+                      <div className="max-w-[250px]">{getCustomFieldDisplayValue(field.value as unknown as ShelfAssetCustomFieldValueType["value"])}</div>
```
Comment (DonKoko) [inline app/routes/_layout+/assets.$assetId.tsx]: Lets align the multiline text to the right.
Screenshot 2023-09-26 at 15 24 10

File: app/components/assets/import-content.tsx [hunk 4]
```diff
@@ -84,7 +84,10 @@ export const ImportContent = () => (
         assets. A new asset will be created for each valid row in the sheet.
       </li>
       <li>
-        To import custom fields prefix, your column heading with <b>"cf: "</b>.
+        To import custom fields, prefix your column heading with <b>"cf: "</b>, add the type followed by a coma from one of the allowed types(<b>"text", "boolean", "option", "multiline text", "date"</b>).
```
Comment (DonKoko) [inline app/components/assets/import-content.tsx:87]: This is a good explanation but I think we need to write a blog post with more details about this. @carlosvirreira 
@mahendraHegde also I am wondering how does it work with options?

**Review by DonKoko — CHANGES_REQUESTED** [review]
I have a few small changes required.

File: app/components/assets/import-content.tsx (hunk 4 above)
Comment (mahendraHegde) [reply app/components/assets/import-content.tsx:87]: for options, we check if the custom field and option both exists, if not we create them, if only option is new, we add the new option to the existing custom field

Comment (DonKoko) [reply app/components/assets/import-content.tsx:87]: Maybe good to also add that to the explanation in you added.

**Comment by DonKoko** [comment]
Tests are failing because I didnt update them to use node-18 yet. I will merge the PR with the tests failing and will then fix it in dev so it can be deployed. Will also be adjusted in main branch.

**Comment by DonKoko** [comment]
@mahendraHegde did you solve the issue with `eslint-plugin-unused-imports`? I still cant install it on node-18.

**Pull request merged**